man/mhstore.man
.TH MHSTORE %manext1% 2015-02-06 "%nmhversion%"
.
.\" %nmhwarning%
.
.SH NAME
mhstore \- store contents of nmh MIME messages into files
.SH SYNOPSIS
.HP 5
.na
.B mhstore
.RB [ \-help ]
.RB [ \-version ]
.RI [ +folder ]
.RI [ msgs ]
.RB [ \-file
.IR file ]
.RB [ \-outfile
.IR outfile ]
.RB [ \-part
.IR number ]
\&...
.RB [ \-type
.IR content ]
\&...
.RB [ \-prefer
.IR content ]
\&...
.RB [ \-noprefer ]
.RB [ \-auto " | " \-noauto ]
.RB [ \-clobber
.IR always " | " auto " | " suffix " | " ask " | " never ]
.RB [ \-verbose " | " \-noverbose ]
.ad
.SH DESCRIPTION
The
.B mhstore
command allows you to store the contents of a collection of MIME
(multi-media) messages into files or other messages.
.PP
.B mhstore
manipulates multi-media messages as specified in RFC 2045 to RFC 2049.
.PP
By default,
.B mhstore
will store all the parts of each message.
Each part will be stored in a separate file.  The header fields of
the message are not stored.  By using the
.BR \-part ,
.BR \-type ,
and
.B \-prefer
switches, you may limit and reorder the set of parts to be stored,
based on part number and/or content type.
.PP
The
.B \-file
.I file
switch directs
.B mhstore
to use the specified file as the source message, rather than a message
from a folder.  If you specify this file as \*(lq-\*(rq, then
.B mhstore
will accept the source message on the standard input.  Note that the
file, or input from standard input, should be a validly formatted
message, just like any other
.B nmh
message.  It should
.I not
be in mail drop format (to convert a file in
mail drop format to a folder of
.B nmh
messages, see
.IR inc (1)).
.PP
A part specification consists of a series of numbers separated by
dots.  For example, in a multipart content containing three parts,
these would be named as 1, 2, and 3, respectively.  If part 2 was also
a multipart content containing two parts, these would be named as 2.1
and 2.2, respectively.  Note that the
.B \-part
switch is effective only for messages containing a multipart content.
If a message has some other kind of content, or if the part is itself
another multipart content, the
.B \-part
switch will not prevent the content from being acted upon.
.PP
The
.B \-type
switch can also be used to restrict (or, when used in conjunction with
.BR \-part ,
to further restrict) the selection of parts according to content type.
One or more
.B \-type
switches part will only select the first match from a
multipart/alternative, even if there is more than one
subpart that matches (one of) the given content type(s).
.PP
Using either
.B \-part
or
.B -type
switches alone will cause either to select the part(s) they match.
Using them together will select only the part(s) matched by both
(sets of) switches.  In other words, the result is the intersection,
and not the union, of their separate match results.
.PP
A content specification consists of a content type and a subtype.
The initial list of \*(lqstandard\*(rq content types and subtypes can
be found in RFC 2046.
.PP
A list of commonly used contents is briefly reproduced here:
.PP
.RS 5
.nf
.ta \w'application  'u
Type	Subtypes
----	--------
text	plain, enriched
multipart	mixed, alternative, digest, parallel
message	rfc822, external-body
application	octet-stream, postscript
image	jpeg, gif, png
audio	basic
video	mpeg
.fi
.RE
.PP
A legal MIME message must contain a subtype specification.
.PP
To specify a content, regardless of its subtype, just use the name
of the content, e.g., \*(lqaudio\*(rq.  To specify a specific
subtype, separate the two with a slash, e.g., \*(lqaudio/basic\*(rq.
Note that regardless of the values given to the
.B \-type
switch, a multipart content (of any subtype listed above) is always acted
upon.  Further note that if the
.B \-type
switch is used, and it is desirable to act on a message/external-body
content, then the
.B \-type
switch must be used twice: once for message/external-body and once for
the content externally referenced.
.PP
The
.B \-prefer
switch will alter the part-ordering of multipart/alternative MIME sections
in order to override the sender-imposed default ordering.
The
.B \-prefer
switch is functionally most important for
.BR mhshow ,
but is also implemented in
.B mhlist
and
.B mhstore
to make common part-numbering possible across all three programs.
The last of multiple
.B \-prefer
switches will have the highest priority.  This allows the command line
switches to override profile entries.
See
.IR mhlist (1)
and
.IR mhshow (1)
for more information on
.BR \-prefer .
.PP
The
.B \-noprefer
switch will cancel any previous
.B \-prefer
switches.
.SS "Storing the Contents"
.B mhstore
will store the contents of the named messages in
\*(lqnative\*(rq (decoded) format.  Two things must be determined:
the directory in which to store the content, and the filenames.
Files are written in the directory given by the
\*(lqnmh-storage\*(rq profile entry, e.g.,
.PP
.RS 5
nmh-storage: /tmp
.RE
.PP
If this entry isn't present, the current working directory is used.
.PP
If the
.B \-outfile
switch is given, its argument is used for the filename to store all
of the content, with \*(lq-\*(rq indicating standard output.  If the
.B \-auto
switch is given, then
.B mhstore
will check if the message contains information indicating the filename
that should be used to store the content.  This information should be
specified as the \*(lqfilename\*(rq attribute in the
\*(lqContent-Disposition\*(rq header or as the \*(lqname\*(rq
attribute in the \*(lqContent-Type\*(rq header for the content you are
storing.  For security reasons, this filename will be ignored if it
begins with the character '/', '.', '|', or '!', or if it contains the
character '%'.  We also recommend using a \*(lqnmh-storage\*(rq profile
entry or a
.B \-clobber
switch setting other than the default of \*(lqalways\*(rq to avoid
overwriting existing files.
.PP
If the
.B \-auto
switch is not given (or is being ignored for security reasons) then
.B mhstore
will look in the user's profile for a \*(lqformatting string\*(rq to
determine how the different contents should be stored.  First,
.B mhstore
will look for an entry of the form:
.PP
.RS 5
mhstore-store-<type>/<subtype>
.RE
.PP
to determine the formatting string.  If this isn't found,
.B mhstore
will look for an entry of the form:
.PP
.RS 5
mhstore-store-<type>
.RE
.PP
to determine the formatting string.
.PP
If the formatting string starts with a \*(lq+\*(rq character, then
content is stored in the named folder.  A formatting string consisting
solely of a \*(lq+\*(rq character is interpreted to be the current
folder.
.PP
If the formatting string consists solely of a \*(lq-\*(rq character,
then the content is sent to the standard output.
.PP
If the formatting string starts with a '|', then it represents
a command for
.B mhstore
to execute which should ultimately store the content.
The content will be passed to the standard input of the command.
Before the command is executed,
.B mhstore
will change to the appropriate directory, and any escapes (given below)
in the formatting string will be expanded.
The use of the \*(lq%a\*(rq sequence is not recommended because
the user has no control over the Content-Type parameter data.
.PP
Otherwise, the formatting string will represent a pathname in which
to store the content.  If the formatting string starts with a '/',
then the content will be stored in the full path given, else the
file name will be relative to the value of \*(lqnmh-storage\*(rq or
the current working directory.  Any escapes (given below) will be
expanded, except for the a-escape.  Note that if \*(lqnmh-storage\*(rq
is not an absolute path, it will be relative to the folder that
contains the message(s).
.PP
A command or pathname formatting string may contain the following
escapes.  If the content isn't part of a multipart (of any subtype
listed above) content, the p-escapes are ignored.
.PP
.RS 5
.nf
.ta \w'%P  'u
%a	Parameters from Content-Type  (only valid with command)
%m	Insert message number
%P	Insert part number with leading dot
%p	Insert part number without leading dot
%t	Insert content type
%s	Insert content subtype
%%	Insert character %
.fi
.RE
.PP
If no formatting string is found,
.B mhstore
will check to see if the content is application/octet-stream with parameter
\*(lqtype=tar\*(rq.  If so,
.B mhstore
will choose an appropriate filename.  If the content is not
application/octet-stream, then
.B mhstore
will check to see if the content is a message.  If so,
.B mhstore
will use the value \*(lq+\*(rq.  As a last resort,
.B mhstore
will use the value \*(lq%m%P.%s\*(rq.
.PP
Example profile entries might be:
.PP
.RS 5
.nf
mhstore-store-text: %m%P.txt
mhstore-store-text: +inbox
mhstore-store-audio/basic: | raw2audio -e ulaw -s 8000 -c 1 > %m%P.au
mhstore-store-image/jpeg: %m%P.jpg
mhstore-store-application/PostScript: %m%P.ps
.fi
.RE
.PP
The
.B \-verbose
switch directs
.B mhstore
to print out the names of files that it stores.  For backward
compatibility, it is the default.  The
.B \-noverbose
switch suppresses these printouts.
.SS "Overwriting Existing Files"
The
.B \-clobber
switch controls whether
.B mhstore
should overwrite existing files.  The allowed values for this switch
and corresponding behavior when
.B mhstore
encounters an existing file are:
.PP
.RS 5
.nf
.ta \w'suffix  'u
always    Overwrite existing file (default)
auto      Create new file of form name-n.extension
suffix    Create new file of form name.extension.n
ask       Prompt the user to specify whether or not to overwrite
          the existing file
never     Do not overwrite existing file
.fi
.RE
.PP
With
.I auto
and
.IR suffix ,
.I n
is the lowest unused number, starting from one, in the same form.  If
a filename does not have an extension (following a '.'), then
.I auto
and
.I suffix
create a new file of the form
.I name-n
and
.IR name.n ,
respectively.  With
.I never
and
.IR ask ,
the exit status of
.B mhstore
will be the number of files that were requested but not stored.
.PP
With
.IR ask ,
if standard input is connected to a terminal, the user is prompted to
respond
.IR yes ,
.IR no ,
or
.IR rename ,
to whether the file should be overwritten.  The responses
can be abbreviated.  If the user responds with
.IR rename ,
then
.B mhstore
prompts the user for the name of the new file to be created.  If it is
a relative path name (does not begin with '/'), then it is relative to
the current directory.  If it is an absolute or relative path to a
directory that does not exist, the user will be prompted whether to
create the directory.  If standard input is not connected to a
terminal,
.I ask
behaves the same as
.IR always .
.SS "External Access"
For contents of type message/external-body,
\fImhstore\fR supports these access-types:
.IP \(bu 4
afs
.IP \(bu 4
anon-ftp
.IP \(bu 4
ftp
.IP \(bu 4
local-file
.IP \(bu 4
mail-server
.IP \(bu 4
url
.PP
For the \*(lqanon-ftp\*(rq and \*(lqftp\*(rq access types,
.B mhstore
will look for the \*(lqnmh-access-ftp\*(rq
profile entry, e.g.,
.PP
.RS 5
nmh-access-ftp: myftp.sh
.RE
.PP
to determine the pathname of a program to perform the FTP retrieval.
This program is invoked with these arguments:
.PP
.RS 5
.nf
domain name of FTP-site
username
password
remote directory
remote filename
local filename
\*(lqascii\*(rq or \*(lqbinary\*(rq
.fi
.RE
.PP
The program should terminate with an exit status of zero if the
retrieval is successful, and a non-zero exit status otherwise.
.PP
For the \*(lqurl\*(rq access types,
.B mhstore
will look for the \*(lqnmh-access-url\*(rq profile entry, e.g.,
.PP
.RS 5
nmh-access-url: curl -L
.RE
.PP
to determine the program to use to perform the HTTP retrieval.  This program
is invoked with one argument: the URL of the content to retrieve.  The program
should write the content to standard out, and should terminate with a status
of zero if the retrieval is successful and a non-zero exit status otherwise.
.SS "User Environment"
Because the environment in which
.B mhstore
operates may vary for different machines,
.B mhstore
will look for the environment variable MHSTORE .
If present, this specifies the name of an additional user profile
which should be read.  Hence, when a user logs in on a particular
machine, this environment variable should be set to refer to a
file containing definitions useful for that machine.  Finally,
.B mhstore
will attempt to consult
.PP
.RS 5
%nmhetcdir%/mhn.defaults
.RE
.PP
which is created automatically during
.B nmh
installation.
.PP
See "Profile Lookup" in
.IR mh-profile (5)
for the profile search order, and for how duplicate entries are treated.
.SH EXAMPLES
.SS Decoding RFC 2047-encoded file names
The improper RFC 2047 encoding of file name parameters can be replaced
with correct RFC 2231 encoding using
.BR mhfixmsg ,
either permanently or ephemerally, e.g.,
.PP
.RS
.nf
mhfixmsg -outfile - | mhstore -auto -clobber ask -file -
.fi
.RE
.PP
The
.BI \-clobber ask
is not necessary, though recommended to avoid silently overwriting an existing
file.
.SH FILES
.B mhstore
looks for additional profile files in multiple locations: absolute
pathnames are accessed directly, tilde expansion is done on usernames,
and files are searched for in the user's
.I Mail
directory as specified in their profile.  If not found there, the directory
.RI \*(lq %nmhetcdir% \*(rq
is checked.
.PP
.fc ^ ~
.nf
.ta \w'%nmhetcdir%/ExtraBigFileName  'u
^$HOME/.mh_profile~^The user profile
^$MHSTORE~^Additional profile entries
^%nmhetcdir%/mhn.defaults~^System default MIME profile entries
.fi
.SH "PROFILE COMPONENTS"
.fc ^ ~
.nf
.ta 2.4i
.ta \w'ExtraBigProfileName  'u
^Path:~^To determine the user's nmh directory
^Current\-Folder:~^To find the default current folder
^nmh-access-ftp:~^Program to retrieve contents via FTP
^nmh-access-url:~^Program to retrieve contents via HTTP
^nmh-storage~^Directory to store contents
^mhstore-store-<type>*~^Template for storing contents
.fi
.SH "SEE ALSO"
.IR mhbuild (1),
.IR mhfixmsg (1),
.IR mhlist (1),
.IR mhshow (1),
.IR sendfiles (1)
.SH DEFAULTS
.nf
.RB ` +folder "' defaults to the current folder"
.RB ` msgs "' defaults to cur"
.RB ` \-noauto '
.RB ` \-clobber\ always '
.RB ` \-nocheck '
.RB ` \-verbose '
.SH CONTEXT
If a folder is given, it will become the current folder.  The last
message selected will become the current message.
.SH BUGS
Partial messages contained within a multipart content are not reassembled.


NEWS
# NEWS - Release notes for nmh 1.7
#

Welcome to nmh, the new version of the classic MH mail handling system.
It's been over three years since the last release of nmh, and there have
been a number of significant changes since the last release.  Long-time
MH and nmh uses should read careful the NOTEABLE CHANGES section, as there
are some significant changes to nmh behavior.  Otherwise, please see the
README and INSTALL files for help on getting started with nmh.

For news of future releases, subscribe to the low-volume
    https://lists.nongnu.org/mailman/listinfo/nmh-announce

---------------
NOTABLE CHANGES
---------------

The largest notable changes in the 1.7 release are:

- Complete unification of network security support.  All network protocols
  (currently, POP and SMTP) have been refactored to use a common set of
  security routines.  This means all protocols support all SASL mechanisms
  (via the Cyrus-SASL library) and TLS.  TLS support has been strengthened
  to perform certificate name validation and to require TLS 1.1 as a
  minimum protocol.  Also, all protocols can make use of the OAuth2/XOAUTH
  SASL mechanism, which is supported by Gmail.
- send(1) now supports adding switches to post(8) based on the address or
  domain of the email address in the From: header; this more easily allows
  users to support multiple identities.
- A generic facility for passing arguments to filter programs in repl(1)
  by use of the -convertargs switch.
- Native support for the manipulation of iCalendar requests; see mhical(1)
  for more details.

------------
NEW FEATURES
------------

The following are new features for the 1.7 release of nmh:

- When building from source, configure will derive ${prefix} from an existing
  nmh installation if it finds one in your $PATH.
- Added welcome message when nmh detects that its version changed.
- The default locations for configuration files and support binaries
  have been changed.  Configuration files now install into ${sysconfdir}/nmh,
  and support binaries are placed in ${libexecdir}/nmh.  If you are upgrading
  an existing installation you should look for old configuration files in
  ${sysconfdir} and merge any local customizations into the new files in
  ${sysconfdir}/nmh, then remove the old files.  ${libdir} will also contain
  obsolete support programs that should be removed.
- All TLS connections now perform certificate validation (including hostname
  matching) by default; can be disabled on a per-application basis.
- post now defaults to port 587 on 'smtp' message submission.
- A value of 0 for the width switch of scan(1), inc(1), ap(1), dp(1),
  fmttest(1), and mhl(1) now means as many characters as the format
  engine can produce [Bug #15274].  That amount is limited by internal
  buffers.
- If a component has trailing whitespace, e.g., body:component="> ",
  mhl now trims that whitespace off when filtering blank text lines.
- An "rtrim" flag has been added to mhl to remove any trailing
  whitespace from filtered text lines.  A corresponding "nortrim" flag
  has also been added.
- Added getmymbox and getmyaddr mh-format(5) function escapes.
- New -[no]changecur, -fixtype, -decodetypes, and -[no]crlflinebreaks switches
  have been added to mhfixmsg(1).
- mhfixmsg now removes an extraneous trailing semicolon from header
  parameter lists.
- Added -convertargs switch to repl(1), to pass arguments to programs
  specified in the user's profile or mhn.defaults to convert message
  content.
- Added mhical(1), to display, reply to, and cancel iCalendar (RFC 5545)
  event requests.
- Added multiply mh-format(5) function.
- "mhparam bindir" prints the path to the directory containing the public
  executables (${bindir}).
- New "-prefer" switch for mhshow (and mhlist and mhshow), to allow specifying
  the preferred content types to show, if present in a multipart alternative.
- mh-format now has %(kilo) and %(kibi) functions, to allow printing
  numbers with SI or IEC quantities, e.g. "10K", "2.3Mi".
- Support for the -sendmail flag to send/post to change the sendmail
  binary when using the sendmail/pipe MTS.
- Added support to send(1) to specify switches to post(1) based on address or
  domain name in From: header line in message draft.
- post(8) -snoop now attempts to decode base64-encoded SMTP traffic.
- folder(1) -nocreate now prints a warning message for a non-existent folder.
- mhfixmsg(1) now allows -decodetext binary, though 8bit is still the default.
- inc(1) and msgchk(1) now support TLS encryption natively.
- All network protocols support the XOAUTH authentication mechanism.
- Support for SMTPUTF8 (RFC 6531) has been added.  mhshow(1) already supported
  RFC 6532, assuming all 8-bit message header field bodies are UTF-8 and use
  of a UTF-8 locale.
- mhfixmsg now replaces RFC 2047 encoding with RFC 2231 encoding of name and
  filename parameters in Content-Type and Content-Disposition headers,
  respectively.
- If a message body contains 8-bit bytes, post(8) uses SMTP 8BITMIME if the
  server supports it.  If not, post fails with a message to the user to
  encode the message for 7-bit transport.
- Fewer lseek(2)s will be used when reading headers in the common case.
- ./configure's --enable-debug has been removed;  it did nothing.
- configure now defaults to enabling each of TLS and Cyrus SASL if the
  necessary headers and libraries are found.
- Moved build_nmh to top-level directory.
- Better error reporting for connections to network services.

---------
BUG FIXES
---------

- The format scanner no longer subtracts 1 from the width.  This has the
  effect of no longer counting the trailing newline in the output of
  scan(1), inc(1), and the other programs that rely on the format scanner.
- The first character of some very short (less than 4 characters) message
  bodies is no longer dropped.
- Single-character headers can be reliably formatted, etc., instead of
  apparently being missing.

uip/mhstore.c
/* mhstore.c -- store the contents of MIME messages
 *
 * This code is Copyright (c) 2002, by the authors of nmh.  See the
 * COPYRIGHT file in the root directory of the nmh distribution for
 * complete copyright information.
 */

#include "h/mh.h"
#include "sbr/m_name.h"
#include "sbr/m_gmprot.h"
#include "sbr/getarguments.h"
#include "sbr/seq_setprev.h"
#include "sbr/seq_setcur.h"
#include "sbr/seq_save.h"
#include "sbr/smatch.h"
#include "sbr/m_convert.h"
#include "sbr/getfolder.h"
#include "sbr/folder_read.h"
#include "sbr/context_save.h"
#include "sbr/context_replace.h"
#include "sbr/context_find.h"
#include "sbr/readconfig.h"
#include "sbr/ambigsw.h"
#include "sbr/path.h"
#include "sbr/print_version.h"
#include "sbr/print_help.h"
#include "sbr/error.h"
#include <fcntl.h>
#include "h/signals.h"
#include "h/mts.h"
#include "h/tws.h"
#include "h/mime.h"
#include "h/mhparse.h"
#include "h/done.h"
#include "h/utils.h"
#include "mhmisc.h"
#include "sbr/m_maildir.h"
#include "mhfree.h"

#define MHSTORE_SWITCHES \
    X("auto", 0, AUTOSW) \
    X("noauto", 0, NAUTOSW) \
    X("check", -5, CHECKSW) \
    X("nocheck", -7, NCHECKSW) \
    X("verbose", 0, VERBSW) \
    X("noverbose", 0, NVERBSW) \
    X("file file", 0, FILESW)		/* interface from show */ \
    X("outfile outfile", 0, OUTFILESW) \
    X("part number", 0, PARTSW) \
    X("type content", 0, TYPESW) \
    X("prefer content", 0, PREFERSW) \
    X("noprefer", 0, NPREFERSW) \
    X("version", 0, VERSIONSW) \
    X("help", 0, HELPSW) \
    X("clobber always|auto|suffix|ask|never", 0, CLOBBERSW) \
    X("debug", -5, DEBUGSW) \

#define X(sw, minchars, id) id,
DEFINE_SWITCH_ENUM(MHSTORE);
#undef X

#define X(sw, minchars, id) { sw, minchars, id },
DEFINE_SWITCH_ARRAY(MHSTORE, switches);
#undef X


#define	quitser	pipeser

/* mhparse.c */
int debugsw = 0;

/*
 * static prototypes
 */
static void pipeser (int);


int
main (int argc, char **argv)
{
    int msgnum;
    bool autosw = false;
    /* verbosw defaults to 1 for backward compatibility. */
    bool verbosw = true;
    const char *clobbersw = "always";
    char *cp, *file = NULL, *outfile = NULL, *folder = NULL;
    char *maildir, buf[100], **argp;
    char **arguments;
    char *cwd;
    struct msgs_array msgs = { 0, 0, NULL };
    struct msgs *mp = NULL;
    CT ct, *ctp;
    FILE *fp;
    int files_not_clobbered;
    mhstoreinfo_t info;

    if (nmh_init(argv[0], true, true)) { return 1; }

    set_done(freects_done);

    arguments = getarguments (invo_name, argc, argv, 1);
    argp = arguments;

    /*
     * Parse arguments
     */
    while ((cp = *argp++)) {
	if (*cp == '-') {
	    switch (smatch (++cp, switches)) {
	    case AMBIGSW:
		ambigsw (cp, switches);
		done (1);
	    case UNKWNSW:
		die("-%s unknown", cp);

	    case HELPSW:
		snprintf (buf, sizeof(buf), "%s [+folder] [msgs] [switches]",
			invo_name);
		print_help (buf, switches, 1);
		done (0);
	    case VERSIONSW:
		print_version(invo_name);
		done (0);

	    case AUTOSW:
		autosw = true;
		continue;
	    case NAUTOSW:
		autosw = false;
		continue;

	    case CHECKSW:
	    case NCHECKSW:
		/* Currently a NOP */
		continue;

	    case PARTSW:
		if (!(cp = *argp++) || *cp == '-')
		    die("missing argument to %s", argp[-2]);
		if (npart >= NPARTS)
		    die("too many parts (starting with %s), %d max",
			   cp, NPARTS);
		parts[npart++] = cp;
		continue;

	    case TYPESW:
		if (!(cp = *argp++) || *cp == '-')
		    die("missing argument to %s", argp[-2]);
		if (ntype >= NTYPES)
		    die("too many types (starting with %s), %d max",
			   cp, NTYPES);
		types[ntype++] = cp;
		continue;

	    case PREFERSW:
		if (!(cp = *argp++) || *cp == '-')
		    die("missing argument to %s", argp[-2]);
		if (npreferred >= NPREFS)
		    die("too many preferred types (starting with %s), %d max",
			   cp, NPREFS);
		mime_preference[npreferred].type = cp;
		cp = strchr(cp, '/');
		if (cp) *cp++ = '\0';
		mime_preference[npreferred++].subtype = cp;
		continue;

	    case NPREFERSW:
		npreferred = 0;
		continue;

	    case FILESW:
		if (!(cp = *argp++) || (*cp == '-' && cp[1]))
		    die("missing argument to %s", argp[-2]);
		file = *cp == '-' ? cp : path (cp, TFILE);
		continue;

	    case OUTFILESW:
		if (!(cp = *argp++) || (*cp == '-' && cp[1]))
		    die("missing argument to %s", argp[-2]);
		outfile = *cp == '-' ? cp : path (cp, TFILE);
		continue;

	    case VERBSW:
		verbosw = true;
		continue;
	    case NVERBSW:
		verbosw = false;
		continue;
            case CLOBBERSW:
		if (!(cp = *argp++) || *cp == '-')
		    die("missing argument to %s", argp[-2]);
		clobbersw = cp;
		continue;
	    case DEBUGSW:
		debugsw = 1;
		continue;
	    }
	}
	if (*cp == '+' || *cp == '@') {
	    if (folder)
		die("only one folder at a time!");
            folder = pluspath (cp);
	} else
		app_msgarg(&msgs, cp);
    }

    /* null terminate the list of acceptable parts/types */
    parts[npart] = NULL;
    types[ntype] = NULL;

    /*
     * Check if we've specified an additional profile
     */
    if ((cp = getenv ("MHSTORE"))) {
	if ((fp = fopen (cp, "r"))) {
	    readconfig(NULL, fp, cp, 0);
	    fclose (fp);
	} else {
	    admonish ("", "unable to read $MHSTORE profile (%s)", cp);
	}
    }

    /*
     * Read the standard profile setup
     */
    if ((fp = fopen (cp = etcpath ("mhn.defaults"), "r"))) {
	readconfig(NULL, fp, cp, 0);
	fclose (fp);
    }

    /*
     * Cache the current directory before we do any chdirs()'s.
     */
    cwd = mh_xstrdup(pwd());

    if (!context_find ("path"))
	free (path ("./", TFOLDER));

    if (file && msgs.size)
	die("cannot specify msg and file at same time!");

    /*
     * check if message is coming from file
     */
    if (file) {
	cts = mh_xcalloc(2, sizeof *cts);
	ctp = cts;

	if ((ct = parse_mime (file))) {
	    *ctp++ = ct;
	    if (outfile) {
		ct->c_storage = mh_xstrdup(outfile);
	    }
        }
    } else {
	/*
	 * message(s) are coming from a folder
	 */
	if (!msgs.size)
	    app_msgarg(&msgs, "cur");
	if (!folder)
	    folder = getfolder (1);
	maildir = m_maildir (folder);

	if (chdir (maildir) == NOTOK)
	    adios (maildir, "unable to change directory to");

	/* read folder and create message structure */
	if (!(mp = folder_read (folder, 1)))
	    die("unable to read folder %s", folder);

	/* check for empty folder */
	if (mp->nummsg == 0)
	    die("no messages in %s", folder);

	/* parse all the message ranges/sequences and set SELECTED */
	for (msgnum = 0; msgnum < msgs.size; msgnum++)
	    if (!m_convert (mp, msgs.msgs[msgnum]))
		done (1);
	seq_setprev (mp);	/* set the previous-sequence */

	cts = mh_xcalloc(mp->numsel + 1, sizeof *cts);
	ctp = cts;

	for (msgnum = mp->lowsel; msgnum <= mp->hghsel; msgnum++) {
	    if (is_selected(mp, msgnum)) {
		char *msgnam;

		msgnam = m_name (msgnum);
		if ((ct = parse_mime (msgnam))) {
		    *ctp++ = ct;
		    if (outfile) {
			ct->c_storage = mh_xstrdup(outfile);
		    }
                }
	    }
	}
    }

    if (!*cts)
	done (1);

    userrs = true;
    SIGNAL (SIGQUIT, quitser);
    SIGNAL (SIGPIPE, pipeser);

    /*
     * Get the associated umask for the relevant contents.
     */
    for (ctp = cts; *ctp; ctp++) {
	struct stat st;

	ct = *ctp;
	if (type_ok (ct, 1) && !ct->c_umask) {
	    if (stat (ct->c_file, &st) != NOTOK)
		ct->c_umask = ~(st.st_mode & 0777);
	    else
		ct->c_umask = ~m_gmprot();
	}
    }

    /*
     * Store the message content
     */
    info = mhstoreinfo_create (cts, cwd, clobbersw, autosw, verbosw);
    store_all_messages (info);
    files_not_clobbered = mhstoreinfo_files_not_clobbered(info);
    mhstoreinfo_free(info);

    /* Now free all the structures for the content */
    for (ctp = cts; *ctp; ctp++)
	free_content (*ctp);

    free (cts);
    cts = NULL;

    /* If reading from a folder, do some updating */
    if (mp) {
	context_replace (pfolder, folder);/* update current folder  */
	seq_setcur (mp, mp->hghsel);	  /* update current message */
	seq_save (mp);			  /* synchronize sequences  */
	context_save ();		  /* save the context file  */
    }

    done (files_not_clobbered);
    return 1;
}


static void
pipeser (int i)
{
    if (i == SIGQUIT) {
	fflush (stdout);
	fprintf (stderr, "\n");
	fflush (stderr);
    }

    done (1);
    /* NOTREACHED */
}
